src/device.ts

```diff
@@ -8,63 +8,100 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
 
-'use strict';
+import {Action, ActionDescription} from './action';
+import Ajv from 'ajv';
+import {Adapter} from './adapter';
+import {Property, PropertyDescription} from './property';
+import {Event, EventDescription} from './event';
 
-const Action = require('./action');
-const Ajv = require('ajv');
 const ajv = new Ajv();
 
-class Device {
-  constructor(adapter, id) {
-    if (typeof id !== 'string') {
-      id = id.toString();
-    }
+export interface DeviceDescription {
+  id: string;
+  title: string;
+  '@context': string;
+  '@type': string[];
+  description: string;
+  properties: Record<string, PropertyDescription>;
+  actions: Record<string, ActionDescription>;
+  events: Record<string, EventDescription>;
+  links: Link[];
+  baseHref?: string;
+  pin: {
+    required: boolean;
+    pattern?: string;
+  };
+  credentialsRequired: boolean;
+}
+
+export interface Link {
+  rel: string;
+  href: string;
+}
+
+export class Device {
+  private adapter: Adapter;
+
+  private id: string;
+
+  private '@context' = 'https://iot.mozilla.org/schemas';
+
+  private '@type': string[] = [];
+
+  private name = '';
+
+  private title = '';
+
+  private description = '';
+
+  private properties = new Map<string, Property<unknown>>();
 
+  private actions = new Map<string, ActionDescription>();
+
+  private events = new Map<string, EventDescription>();
+
+  private links: Link[] = [];
+
+  private baseHref?: string;
+
+  private pinRequired = false;
+
+  private pinPattern?: string;
+
+  private credentialsRequired = false;
+
+  constructor(adapter: Adapter, id: string) {
     this.adapter = adapter;
-    this.id = id;
-    this['@context'] = 'https://iot.mozilla.org/schemas';
-    this['@type'] = [];
-    this.title = '';
-    this.description = '';
-    this.properties = new Map();
-    this.actions = new Map();
-    this.events = new Map();
-    this.links = [];
-    this.baseHref = null;
-    this.pinRequired = false;
-    this.pinPattern = null;
-    this.credentialsRequired = false;
+    this.id = `${id}`;
   }
 
-  asDict() {
-    const properties = {};
-    this.properties.forEach((property, propertyName) => {
-      properties[propertyName] = property.asDict();
-    });
-
-    const actions = {};
-    this.actions.forEach((metadata, actionName) => {
-      actions[actionName] = Object.assign({}, metadata);
+  mapToDict<V>(map: Map<string, V>): Record<string, V> {
+    const dict: Record<string, V> = {};
+    map.forEach((property, propertyName) => {
+      dict[propertyName] = property;
```

Comment:
Before, we were (attempting to) copy the object via `Object.assign({}, metadata)`. Can we keep that behavior to avoid any potential breakage?

Reply:
Good point!
It's always better to avoid side-effects.